monkey/infection_monkey/control.py

```diff
@@ -63,37 +67,66 @@ def wakeup(self, parent=None):
             timeout=MEDIUM_REQUEST_TIMEOUT,
         )
 
-    def find_server(self, default_tunnel=None):
-        logger.debug(f"Trying to wake up with Monkey Island server: {self.server_address}")
-        if default_tunnel:
-            logger.debug("default_tunnel: %s" % (default_tunnel,))
+    def find_server(self, servers: Sequence[str]):
+        logger.debug(f"Trying to wake up with servers: {', '.join(servers)}")
 
-        try:
-            debug_message = "Trying to connect to server: %s" % self.server_address
-            if self.proxies:
-                debug_message += " through proxies: %s" % self.proxies
-            logger.debug(debug_message)
-            requests.get(  # noqa: DUO123
-                f"https://{self.server_address}/api?action=is-up",
-                verify=False,
-                proxies=self.proxies,
-                timeout=MEDIUM_REQUEST_TIMEOUT,
-            )
-            return True
-        except ConnectionError as exc:
-            logger.warning("Error connecting to control server %s: %s", self.server_address, exc)
+        while servers:
+            server = servers[0]
 
-        if self.proxies:
-            return False
-        else:
-            logger.info("Starting tunnel lookup...")
-            proxy_find = tunnel.find_tunnel(default=default_tunnel)
-            if proxy_find:
-                self.set_proxies(proxy_find)
-                return self.find_server()
-            else:
-                logger.info("No tunnel found")
-                return False
+            try:
+                debug_message = f"Trying to connect to server: {server}"
+                logger.debug(debug_message)
+                requests.get(  # noqa: DUO123
+                    f"https://{server}/api?action=is-up",
+                    verify=False,
+                    timeout=MEDIUM_REQUEST_TIMEOUT,
+                )
+
+                # We remove the server that has been succesfull
+                servers.remove(server)
+
+                # TODO: Check how we are going to set the server address that the ControlCLient
+                # is going to use
+                # self.server_address = server
+
+                # If we have any other server we send them RELAY_CONTROL_MESSAGE
+                if servers:
```

Comment:
Why are we doing this? So that agents won't have to wait for children connection?

Reply:
Yes. They can then shut down if they're not expecting to a connection from any children.